cdap-app-fabric/src/main/java/co/cask/cdap/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,177 @@
+  }));
+
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final Type MAP_STRING_MESSAGE_TYPE = new TypeToken<Map<String, List<MonitorMessage>>>() { }.getType();
+
+  private final RESTClient restClient;
+  private final ClientConfig clientConfig;
+
+  private final int limit;
+  private final String runtimeId;
+  private final CConfiguration cConf;
+  private Map<String, MonitorConsumeRequest> topicsToRequest;
+
+  private final MessagePublisher messagePublisher;
+  private final long pollTimeMillis;
+  private volatile Thread runThread;
+  private volatile boolean stopped;
+
+  public RuntimeMonitor(String runtimeId, CConfiguration cConf, MessagePublisher messagePublisher,
+                        ClientConfig clientConfig) {
+    this.runtimeId = runtimeId;
+    this.cConf = cConf;
+    this.messagePublisher = messagePublisher;
+    this.clientConfig = clientConfig;
+    this.restClient = new RESTClient(clientConfig);
+    this.limit = cConf.getInt(Constants.RuntimeMonitor.RUNTIME_MONITOR_BATCH_LIMIT);
+    this.pollTimeMillis = cConf.getInt(Constants.RuntimeMonitor.RUNTIME_MONITOR_POLLTIME_MS);
+    this.topicsToRequest = new HashMap<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.debug("Starting runtime monitor {}....", runtimeId);
+    initializeTopics();
+  }
+
+  private void initializeTopics() throws Exception {
+    Set<String> topicsToMonitor = new HashSet<>();
+    topicsToMonitor.addAll(Arrays.asList(cConf.get(Constants.RuntimeMonitor.RUNTIME_MONITOR_TOPICS).split(",")));
+
+    HttpResponse response = restClient.execute(HttpMethod.POST, clientConfig.resolveURL("runtime/monitor/topics"),
+                                               GSON.toJson(topicsToMonitor), Collections.emptyMap(), null);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
+      throw new ServiceUnavailableException(response.getResponseBodyAsString());
+    }
+
+    Map<String, String> configToTopic = GSON.fromJson(response.getResponseBodyAsString(Charset.forName("UTF-8")),
+                                                      MAP_STRING_STRING_TYPE);
+
+    // TODO initialize from offset table for a given runtimeId
+    for (Map.Entry<String, String> entry : configToTopic.entrySet()) {
+      topicsToRequest.put(entry.getKey(), new MonitorConsumeRequest(entry.getValue(), entry.getKey(), null, limit));
+    }
+  }
+
+  @Override
+  protected void run() throws Exception {
+    runThread = Thread.currentThread();
+
+    try {
+      while (!stopped) {
+        try {
+          HttpResponse response = restClient.execute(HttpMethod.POST, clientConfig.resolveURL("runtime/metadata"),
+                                                     GSON.toJson(topicsToRequest), Collections.emptyMap(), null);
+
+          if (response.getResponseCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
```

Comment:
BTW, I notice that the retry inside the `RESTClient` always sleep for 1 second  between retries and is not configurable. It might be good to change ClientConfig use `RetryStrategy` instead. That can be a separate PR, but it's good to make that change.

Reply:
I can add that in the same PR.